src/storage/txn/scheduler.rs

```diff
@@ -221,57 +200,169 @@ impl<E: Engine> Scheduler<E> {
             TaskContext::new(lock, callback, cmd)
         };
 
-        self.running_write_bytes += tctx.write_bytes;
-        SCHED_WRITING_BYTES_GAUGE.set(self.running_write_bytes as i64);
-
-        if self.pending_tasks.insert(cid, task).is_some() {
+        let running_write_bytes = self
+            .running_write_bytes
+            .fetch_add(tctx.write_bytes, Ordering::AcqRel) as i64;
+        SCHED_WRITING_BYTES_GAUGE.set(running_write_bytes + tctx.write_bytes as i64);
+
+        if self.pending_tasks[id_index(cid)]
+            .lock()
+            .unwrap()
+            .insert(cid, task)
+            .is_some()
+        {
             panic!("command cid={} shouldn't exist", cid);
         }
         SCHED_CONTEX_GAUGE.set(self.pending_tasks.len() as i64);
-        if self.task_contexts.insert(cid, tctx).is_some() {
+        if self.task_contexts[id_index(cid)]
+            .lock()
+            .unwrap()
+            .insert(cid, tctx)
+            .is_some()
+        {
             panic!("TaskContext cid={} shouldn't exist", cid);
         }
     }
 
-    fn dequeue_task_context(&mut self, cid: u64) -> TaskContext {
-        let tctx = self.task_contexts.remove(&cid).unwrap();
-
-        self.running_write_bytes -= tctx.write_bytes;
-        SCHED_WRITING_BYTES_GAUGE.set(self.running_write_bytes as i64);
+    fn dequeue_task_context(&self, cid: u64) -> TaskContext {
+        let tctx = self.task_contexts[id_index(cid)]
+            .lock()
+            .unwrap()
+            .remove(&cid)
+            .unwrap();
+
+        let running_write_bytes = self
+            .running_write_bytes
+            .fetch_sub(tctx.write_bytes, Ordering::AcqRel) as i64;
+        SCHED_WRITING_BYTES_GAUGE.set(running_write_bytes - tctx.write_bytes as i64);
         SCHED_CONTEX_GAUGE.set(self.pending_tasks.len() as i64);
 
         tctx
     }
 
-    pub fn fetch_executor(&self, priority: CommandPri) -> Executor<E> {
-        let pool = match priority {
-            CommandPri::Low | CommandPri::Normal => &self.worker_pool,
-            CommandPri::High => &self.high_priority_pool,
-        };
-        let pool_scheduler = pool.scheduler();
-        let scheduler = self.scheduler.clone();
-        Executor::new(scheduler, pool_scheduler)
+    fn too_busy(&self) -> bool {
+        fail_point!("txn_scheduler_busy", |_| true);
+        self.running_write_bytes.load(Ordering::Acquire)
+            >= self.sched_pending_write_threshold.load(Ordering::Acquire)
```

Comment:
do we need to define the atomic type for sched_pending_write_threshold?

Reply:
You leave a `TODO` here. I thought you want to introduce a self-adaptable strategy. For that, atomic is necessary.